Update AcrylicBrushes and add new AcrylicBrushes

chingucoding opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

This PR updates the existing SystemControlTransientBackgroundBrush and adds new Acrylic- and SolidColorBrushes.

In addition to that, NavigationView and CommandbarFlyout have been updated to use the new brushes.

~In the issues it says that the SystemControlDefaultBrighteningBrush should have a opacity of `0.**0**419`, however that is almost invisible to see, so I assumed that this was a mistake in the proposal and should have been `0.419`.~

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
<!--- Use the syntax "Closes #1234" or "Fixes #5678" so that GitHub will close the issue once the PR is complete. -->
Closes #3478
## How Has This Been Tested?
<!--- Please describe how you tested your changes. -->
Tested manually
## Screenshots (if appropriate):
<!--- If you are making visual changes to a Control or adding a new Control, please include screenshots showing the result. -->

NavigationView DisplayMode top:

![Image showing NavigationView in top mode, light theme](https://user-images.githubusercontent.com/16122379/97292581-b5676c80-184b-11eb-8f17-98f8986ffea4.png)

![Image showing NavigationView in top mode, dark theme](https://user-images.githubusercontent.com/16122379/97292550-aaacd780-184b-11eb-96d8-6f91eedce015.png)

NavigationView PaneFlyout:

![Image showing NavigationView paneflyout, light theme](https://user-images.githubusercontent.com/16122379/97292641-ce701d80-184b-11eb-8a2b-16bc5c25c451.png)

![Image showing NavigationView paneflyout, light theme](https://user-images.githubusercontent.com/16122379/97292700-dd56d000-184b-11eb-949b-d3e1e6a50445.png)

CommandBarFlyout:

![CommandBarFlyout in light theme](https://user-images.githubusercontent.com/16122379/97292957-3c1c4980-184c-11eb-930f-05a1829d0d96.png)

![CommandBarFlyout in dark theme](https://user-images.githubusercontent.com/16122379/97293006-50604680-184c-11eb-83cf-b49769658261.png)

YuliKl commented:
> In the issues it says that the SystemControlDefaultBrighteningBrush should have a opacity of 0.**0**419, however that is almost invisible to see, so I assumed that this was a mistake in the proposal and should have been 0.419.

No, that wasn't a typo. This is a translucent white layer that we're adding over dark. Even at 4.19% opacity, the white still shows up. The brightening effect is intended to be subtle.

YuliKl commented on dev/Materials/Acrylic/AcrylicBrush_19h1_themeresources.xaml:
```diff
@@ -0,0 +1,80 @@
+        </ResourceDictionary>
+
+        <ResourceDictionary x:Key="HighContrast">
+            <StaticResource x:Key="SystemControlTransientBackgroundBrush" ResourceKey="SystemChromeMediumLowColor" />
+            <StaticResource x:Key="SystemControlTransientAcrylicElementBrush" ResourceKey="SystemChromeMediumLowColor" />
+            <StaticResource x:Key="SystemControlTransientBackgroundInverseBrush" ResourceKey="SystemBaseMediumHighColor" />
+            <StaticResource x:Key="SystemControlBaseAcrylicBrush" ResourceKey="SystemChromeMediumLowColor" />
```
@teaP this is a question for you. I see that in our existing RS2 file, we set
```xaml
<StaticResource x:Key="SystemControlTransientBackgroundBrush" ResourceKey="SystemChromeMediumLowColor" />
```
This surprised me, as I expected a direct reference to one of the 8 HC colors, instead of added indirection. Should these HC definitions follow the other pattern in the same RS2 file?
```xaml
<SolidColorBrush x:Key="SystemControlAltHighAcrylicElementBrush" Color="{ThemeResource SystemColorWindowColor}" />
```

YuliKl replied:
Based on our conversation earlier today, I think we want to be as direct as possible, so reference the HC colors directly instead of going through SystemChromeMediumLowColor or similar.

chingucoding replied:
I've updated the all resources modified by the feature proposal and also updated the `SystemControlTransientBackgroundBrush` as it also didn't reference a HC brush.

YuliKl commented on dev/Materials/Acrylic/AcrylicBrush_rs1_themeresources.xaml:
```diff
@@ -46,6 +46,10 @@
+            <StaticResource x:Key="SystemControlTransientAcrylicElementBrush" ResourceKey="SystemChromeMediumLowColor" />
+            <SolidColorBrush x:Key="SystemControlTransientBackgroundInverseBrush" Color="#F2F2F2" />
+            <StaticResource x:Key="SystemControlBaseAcrylicBrush" ResourceKey="SystemChromeMediumLowColor" />
+            <SolidColorBrush x:Key="SystemControlDefaultBrighteningBrush" Color="#FFFFFF" Opacity="0.419"/>
```
Maybe we don't need this resource redefined in the RS1 file? It's a SolidColorBrush to begin with and shouldn't have any back compat issues.
(Again, looking to @teaP as the authority here. Just my thought.)

chingucoding replied:
I added it to every file as I was not sure how the resource dictionary actually get merged and if it might get dropped if only specified in one dictionary.

YuliKl commented on dev/NavigationView/NavigationView_rs1_themeresources.xaml:
```diff
@@ -12,9 +12,9 @@
     <ResourceDictionary.ThemeDictionaries>
         <ResourceDictionary x:Key="Dark">
-            <StaticResource x:Key="NavigationViewDefaultPaneBackground" ResourceKey="SystemChromeMediumColor" />
+            <StaticResource x:Key="NavigationViewDefaultPaneBackground" ResourceKey="SystemControlTransientAcrylicElementBrush" />
```
I don't know whether we need to touch this file. Feels like this is providing twice the opportunity to avoid crashing down level...which may not be necessary? Or maybe we should define these brushes to use the acrylic resources in the RS1 dictionary (with the understanding that the resources will gracefully fall back to solid colors) and remove their definitions from the NavigationView RS2 dictionary?

chingucoding replied:
Good point, since we are providing the same resource for RS1 and higher, defining it only in the RS1 file should be fine.

teaP replied:
Techincally, WinUI is only RS2 and greater now. So.... I'm actually not sure what happens to these files.

chingucoding commented:
> No, that wasn't a typo. This is a translucent white layer that we're adding over dark. Even at 4.19% opacity, the white still shows up. The brightening effect is intended to be subtle.

Oh, I am sorry about my mistake here. I've updated it to be 4.19%.

ranjeshj commented:
/azp run

chingucoding commented:
I think test failure was CI related, see https://github.com/microsoft/microsoft-ui-xaml/pull/3516#issuecomment-720694022

ranjeshj commented:
/azp run

chingucoding commented:
The issue is a crash within Windows.UI.Xaml.dll:

```
A crash with exception code 0xC000027B occurred in module "Windows.UI.Xaml.dll" in process "MUXControlsTestApp.exe" (pid:108).
```

Looking at the existing issues for that bug, it seems that that is a framework bug: #716 #1712 #3498 #2669

What surprises me a bit is that this only occurring now. Could it be that it is failing to reference resources? Though then, the question is why it is not failing on other versions or builds.

teaP reviewed (changes requested):
Mostly looks good, I would just make changes in rs2+ and not the rs1 files.

chingucoding commented:
Reverted my changes to the NavigationView RS1 file and added changes to the NavigationView RS2 file.

teaP commented:
/azp run

teaP commented:
I guess I just don't quite believe this broke THAT many tests, so I'm going to run the pipeline again; if we get the same result then I'll investigate. :)

teaP commented:
/azp run

ranjeshj commented:
@chingucoding looks like some issue with the 19h1 run. Can you merge with master ? I can kick off a clean run.

chingucoding commented:
Merged latest changes of master into this PR.

ranjeshj commented:
/azp run

ranjeshj commented:
Looks like the 19h1 run has a lot of failures. @kmahone @StephenLPeters  @teaP  as FYI

kmahone commented:
The tests are all failing on 19H1 x86fre. I can reproduce the issue by running the x86fre MUXControlsTestApp on my machine. It hits a crash at startup while looking up the resource "SystemControlTransientBackgroundBrush". It is unclear what is the cause of the failure. That would require further debugging.

chingucoding commented:
Looking at the error log, it seems that the XAML Compiler is unable to create an `Windows.Foundation.IReference<Double>`from the string "0.96" in x86 release configuration. Funnily this PR seems to be the first time where we are setting the value from XAML which might also explain why we are hitting this just now.

@MikeHillberg @ranjeshj Do you have any idea why this might happen or what we could do to mitigate this beside changing the type to Double to prevent that conversion?

jevansaks commented:
I remember this being a problem that .xaml markup can't do nullable types. @rbrid you ran into this with some of the APIs you were designing for Scroller, did you come up with any satisfactory workarounds?

chingucoding commented:
That's very interesting @jevansaks ! What is a bit surprising is the fact that debug x64 seems to be fine while the release x86 is crashing in that case.

jevansaks commented:
Perhaps the Debug version is using the reflection-based XamlMetadataProvider but the release one isn't?

chingucoding commented:
> Perhaps the Debug version is using the reflection-based XamlMetadataProvider but the release one isn't?

This might probably the reason why this fails. The big question now is how to handle this. We could do something in code behind to set that value but that would definitely not be ideal.

jevansaks commented:
2 possibilities:
1) The safest thing would be to just construct these resources in code-behind and add them to the dictionary manually in code-behind in XamlControlsResources.cpp.

-OR-

2) We might able to come up with a WinUI-level workaround....

Perhaps you can try to reproduce the problem in the Debug build by adding this to the application? (per #1521):

```
  <PropertyGroup>
    <EnableTypeInfoReflection>false</EnableTypeInfoReflection>
  </PropertyGroup>
```

If that repros it then I would suggest you debug through the app's generated XamlMetadataProvider and look at the GetXamlType callback. When it gets called for ```IReference`1<Double>``` (or whatever, i can't remember exactly) then see why it fumbles the conversion. We might be able to work around it by providing a handling of it in WinUI`s own IXamlMetadataProvider implementation (which is also codegen, but a code generator that we wrote ourselves).

chingucoding commented:
Since I can repro this in debug builds by disabling `EnableTypeInfoReflection`. Updating our XAMLMetaDataProvider implementation is probably the better solution now and even more in the long term. Will look into this.

chingucoding commented:
I added the following to the XamlMetadataProviderGenerated.tt (showing generated output) and also added a custom XamlType, however that still resulted in the crash as it fails to create a nullable double from XAML.

XamlType:
```c++
class ReferenceDoubleXamlType : public XamlTypeBase {
public:
    ReferenceDoubleXamlType(wstring_view const& typeName)
    {
        m_typeName = typeName;
        m_createFromString = [](const winrt::hstring& param)
        {
            winrt::IReference<double> returnValue;
            return returnValue;
        };
    }

    winrt::IInspectable CreateFromStringInternal(winrt::hstring const& param)
    {
        return box_value(0);
    }
};
```

Type handling (added more then needed just to be sure):
```c++
XamlMetadataProvider::RegisterXamlType(
    L"Windows.Foundation.IReference`1<Double>",
    []() { return winrt::make<ReferenceDoubleXamlType>((PCWSTR)L"Windows.Foundation.IReference`1<Double>"); });
XamlMetadataProvider::RegisterXamlType(
    L"Windows.Foundation.IReference<Double>",
    []() { return winrt::make<ReferenceDoubleXamlType>((PCWSTR)L"Windows.Foundation.IReference<Double>"); });
XamlMetadataProvider::RegisterXamlType(
    L"System.Nullable`1<Double>",
    []() { return winrt::make<ReferenceDoubleXamlType>((PCWSTR)L"System.Nullable`1<Double>"); });
```

I'm still a bit confused where it is getting it's XamlType for the nullable bool from, shouldn't our XamlMetaDataProvider interfere there?

jevansaks commented:
Oops I thought I replied, sorry. When you debugged it, what did you see happening with the XamlMetadataProvider for the double?-typed property value? We would need to understand that first before figuring out how to work around it.

chingucoding commented:
Calling XamlMetadataProvider::GetXamlType for the double? property yields null. For testing purposes, I tried to add a special case there and return a custom XamlType for this but that didn't seem to help either.

jevansaks commented:
I just tried this in a blank UWP app and I'm able to use AcrylicBrush.TintLuminosityOpacity in markup (even with the metadata provider disabled). I also notice that the XamlTypeInfo.g.cs has a different output for the System.Nullable`1 sections in CreateXamlType. I suspect this was fixed in XamlCompiler and we just need to target the newer SDK.

Can you try retargeting the WinUI solution to 19041? (basically search the codebase for 18362 and replace with 19041 in almost all the places)

chingucoding commented:
I tried setting the target SDK version to 19041 but it didn't seem to help. Though maybe I did something wrong in setting the targeted SDK version? I added a new entry to the SDKVersion.props file and updated the file to export that new version, the test apps all seem to reference said resouce.

jevansaks commented:
I asked @RealTommyKlein and he said it's because of the MinVersion of the app, there was a new API added that enabled them to fix this bug in the Xaml Compiler but it only works in 17763+ OS XAML. Unfortunately we can't require our customers to change their min version so we'll have to find some other workaround. I'm not sure there's too much we can do to change this, so ideas off the top of my head:

1) Create the brush instances from code behind and manually add them to the dictionary.
2) Add another property to AcrylicBrush that's not nullable for setting this from markup for downlevel.
3) ???

chingucoding commented:
Given that this wouldn't be a problem with WinUI 3 and WInUI 3 won't allow us to remove existing APIs, I think modifying API is probably not the best idea. Of course creating them from code behind won't scale well. Especially since we need to create theme resources which makes that even more difficult to achieve. Another option we could do is have them be created in XAML but leave out the tint luminosity opacity effect. When loading said themeresource, we then set the property circumventing the XAML creation issue. However, I am not sure how we would be able to do that for both dark and light theme.

jevansaks commented:
@ranjeshj @YuliKl can you think about other options? 

I don't think a new API is terrible -- we don't necessarily have to carry it forward into WinUI 3 since TintLuminosityOpacity works fine there. There's also an outside chance that someone more familiar with the codegen options can sit and come up with a workaround ... I just don't know if it's worth investing in that.

chingucoding commented:
That's true, we wouldn't have to carry it into WinUI 3. How would such API work though? What if both values have been set? I guess changing the existing TintLuminosityOpciaty property to just a double won't work as that would be a breaking change? Having two APis for the same thing which would compete about their relevance isn't great.

ranjeshj commented:
> Given that this wouldn't be a problem with WinUI 3 and WInUI 3 won't allow us to remove existing APIs, I think modifying API is probably not the best idea. Of course creating them from code behind won't scale well. Especially since we need to create theme resources which makes that even more difficult to achieve. Another option we could do is have them be created in XAML but leave out the tint luminosity opacity effect. When loading said themeresource, we then set the property circumventing the XAML creation issue. However, I am not sure how we would be able to do that for both dark and light theme.

Taking the hack a little further. Could the values also be theme resources ? So when the brush is loaded, lookup a themeresource for the value and set it in code. Since the value is also coming from the theme resource, you should get the correct one for each theme...

chingucoding commented:
@rnjeshj @jevansaks I was able to add a workaround in our XAMLControlResources dictionary. After loading it, if we are in 19H1,we load the default and light theme resource and grab references for the acrylicbrushes to change their `TintLuminosityOpacity`. That way, themeresources are updated, the properties are set and we don't rely on newer platform bug fixes while avoiding to introduce a new API.

jevansaks commented on dev/dll/XamlControlsResources.cpp:
```diff
@@ -87,6 +92,61 @@ void XamlControlsResources::UpdateSource()
+    // The MUXC AcrylicBrush's TintLuminosityOpacity is a nullable double though and needs to be set.
+    // Solution: Load theme resources and edit brushes manually.
+    // Since something must go horribly wrong for those lookups to fail, we just assume they exist.
+    if (SharedHelpers::Is19H1OrHigher())
```
Why is there a 19H1+ condition?

chingucoding replied:
TintLuminosityOpacity is only present on 19H1+ and we also have a separate themeresources file because of that. To not interfere with the RS5- brushes and save performance, we are only updating the brushes when on 19H1+. Of course, we could run it in every case as the MUXC brushes guard against it but we don't need to.

ranjeshj replied:
can you add a comment in the 19h1 theme resource file that tint opacity is being set in code ?

chingucoding replied:
Good idea, added a comment for that. Can you rerun CI @ranjeshj ?

ranjeshj commented:
/azp run

ranjeshj commented:
@jevansaks @licanhua @teaP @kmahone If we don't support RS1 anymore, can we bulk remove those files in a separate PR ?

chingucoding commented:
> @jevansaks @licanhua @teaP @kmahone If we don't support RS1 anymore, can we bulk remove those files in a separate PR ?

This is something that was tracked by #1375 (closed now) and more generally #975. Of course we could remove those files sooner than WinUI 3.

ranjeshj commented:
/azp run

mdtauk commented:
> > In the issues it says that the SystemControlDefaultBrighteningBrush should have a opacity of 0.**0**419, however that is almost invisible to see, so I assumed that this was a mistake in the proposal and should have been 0.419.
> 
> No, that wasn't a typo. This is a translucent white layer that we're adding over dark. Even at 4.19% opacity, the white still shows up. The brightening effect is intended to be subtle.

I am curious @YuliKl where will **SystemControlDefaultBrighteningBrush** be used.  Would I be right in thinking that this may be used for "elevated" surfaces so elements stand out from the dark backgrounds?
